lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -146,92 +144,105 @@ public void OnNext(IAllocatedEvaluator allocatedEvaluator)
+            _activeContexts.Add(value);
 
-            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            if (_activeContexts.Count < _dataSet.Count + 1)
             {
-                _reachedUpdateTaskActiveContext = true;
-                RequestMapEvaluators(_dataSet.Count);
-
-                var partialTaskConf =
-                    TangFactory.GetTang()
-                        .NewConfigurationBuilder(new[]
-                        {
-                            TaskConfiguration.ConfigurationModule
-                                .Set(TaskConfiguration.Identifier,
-                                    IMRUConstants.UpdateTaskName)
-                                .Set(TaskConfiguration.Task,
-                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
-                                .Build(),
-                            _configurationManager.UpdateFunctionConfiguration,
-                            _configurationManager.ResultHandlerConfiguration
-                        })
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
+                return;
+            }
 
-                try
-                {
-                    TangFactory.GetTang()
-                        .NewInjector(partialTaskConf, _configurationManager.UpdateFunctionCodecsConfiguration)
-                        .GetInstance<IIMRUResultHandler<TResult>>();
-                }
-                catch (InjectionException)
-                {
-                    partialTaskConf = TangFactory.GetTang().NewConfigurationBuilder(partialTaskConf)
-                        .BindImplementation(GenericType<IIMRUResultHandler<TResult>>.Class,
-                            GenericType<DefaultResultHandler<TResult>>.Class)
-                        .Build();
-                    Logger.Log(Level.Warning,
-                        "User has not given any way to handle IMRU result, defaulting to ignoring it");
-                }
+            AddGroupCommunicationOperators();
+            _groupCommTaskStarter = new TaskStarter(_groupCommDriver, _dataSet.Count + 1);
 
-                _commGroup.AddTask(IMRUConstants.UpdateTaskName);
-                _groupCommTaskStarter.QueueTask(partialTaskConf, activeContext);
-            }
-            else
+            foreach (var activeContext in _activeContexts)
             {
-                string taskId;
-
-                if (!_taskIdStack.TryPop(out taskId))
+                if (_groupCommDriver.IsMasterTaskContext(activeContext))
                 {
-                    Logger.Log(Level.Warning, "No task Ids exist for the active context {0}. Disposing the context.",
-                        activeContext.Id);
-                    activeContext.Dispose();
-                    return;
+                    _reachedUpdateTaskActiveContext = true;
+                    RequestMapEvaluators(_dataSet.Count);
```

Comment:
The same change is in BroadCastandReduce and PipeLineBroadCastandReduce samples and functional tests passed. We definitely need to add functional test for IMRU driver otherwise it would be very dangerous for people change here there every day.